blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -285,47 +285,71 @@ impl Protocoler {
+            })?;
+
+        let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
+
+        let mut results = Vec::new();
+        for deployed_protocol in protocols {
+            let empty_label_hash = deployed_protocol
+                .protocol
+                .info
+                .protocol_specific
+                .empty_label_hash();
+
+            let domain_name = DomainName::new(name_with_tld, empty_label_hash)?;
+            results.push(DomainNameOnProtocol {
+                inner: domain_name,
+                deployed_protocol,
+            });
+        }
+
+        Ok(results)
+    }
+
     pub fn names_options_in_network(
         &self,
         name: &str,
         network_id: i64,
         maybe_filter: Option<NonEmpty<String>>,
     ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
-        let tlds = self
-            .networks
-            .get(&network_id)
-            .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
-            .use_protocols
-            .iter()
-            .filter_map(|protocol_name| {
-                self.protocols
-                    .get(protocol_name)
-                    .map(|protocol| protocol.info.tld_list.iter().cloned())
-            })
-            .flatten()
-            .collect::<Vec<Tld>>();
-
         if name.contains('.') {
-            let direct = self.names_options_in_network_exact(name, network_id, maybe_filter)?;
+            let direct = self.fetch_domain_options(name, network_id, maybe_filter)?;
             if direct.is_empty() {
                 Err(ProtocolError::InvalidName {
                     name: name.to_string(),
                     reason: "No matching protocols for given TLD".to_string(),
                 })
             } else {
-                Ok(direct.into_iter().take(MAX_NAMES_LIMIT).collect())
+                Ok(direct.into_iter().take(1).collect())
             }
         } else {
+            let tlds = self
+                .networks
+                .get(&network_id)
+                .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
+                .use_protocols
+                .iter()
+                .filter_map(|protocol_name| self.protocols.get(protocol_name))
+                .flat_map(|protocol| protocol.info.tld_list.iter().cloned())
+                .collect::<Vec<Tld>>();
+
             let all_names_with_protocols: Vec<_> = tlds
                 .into_iter()
                 .map(|tld| format!("{}.{}", name, tld.0))
                 .flat_map(|name_with_tld| {
-                    self.names_options_in_network_with_suggestions(
-                        &name_with_tld,
-                        network_id,
-                        maybe_filter.clone(),
-                    )
-                    .unwrap_or_default()
+                    self.fetch_domain_options(&name_with_tld, network_id, maybe_filter.clone())
+                        .unwrap_or_default()
```

Comment:
Does it make sense to log an error or all `Err(_)` variants represent a normal program flow?

Reply:
nice deduplication btw